src/internet/radiobrowser/radiobrowserservice.cpp

```diff
@@ -0,0 +1,390 @@
+    item->setData(branch.listUrl, RadioBrowserService::Role_ListUrl);
+    item->setData(branch.itemsUrl, RadioBrowserService::Role_ItemsUrl);
+    item->setData(true, InternetModel::Role_CanLazyLoad);
+    root_->appendRow(item);
+  }
+}
+
+void RadioBrowserService::RefreshCategory(QStandardItem* item) {
+  QString determinedUrl = item->data(RadioBrowserService::Role_ListUrl)
+                              .toString()
+                              .arg(main_server_url_);
+  QUrl url(determinedUrl);
+
+  QNetworkReply* reply = network_->get(QNetworkRequest(url));
+  int task_id = app_->task_manager()->StartTask(tr("Getting channels"));
+
+  NewClosure(reply, SIGNAL(finished()), this,
+             SLOT(RefreshCategoryFinished(QNetworkReply*, int, QStandardItem*)),
+             reply, task_id, item);
+}
+
+void RadioBrowserService::RefreshCategoryItem(QStandardItem* item) {
+  QStandardItem* parent = item->parent();
+  QString determinedUrl = parent->data(RadioBrowserService::Role_ItemsUrl)
+                              .toString()
+                              .arg(main_server_url_, item->text());
+  QUrl url(determinedUrl);
+
+  QNetworkReply* reply = network_->get(QNetworkRequest(url));
+  int task_id = app_->task_manager()->StartTask(tr("Getting channels"));
+
+  NewClosure(reply, SIGNAL(finished()), this,
+             SLOT(RefreshStreamsFinished(QNetworkReply*, int, QStandardItem*)),
+             reply, task_id, item);
+}
+
+void RadioBrowserService::RefreshTop100(QStandardItem* item) {
+  QString determinedUrl = item->data(RadioBrowserService::Role_ItemsUrl)
+                              .toString()
+                              .arg(main_server_url_);
+  QUrl url(determinedUrl);
+
+  QNetworkReply* reply = network_->get(QNetworkRequest(url));
+  int task_id = app_->task_manager()->StartTask(tr("Getting channels"));
+
+  NewClosure(reply, SIGNAL(finished()), this,
+             SLOT(RefreshStreamsFinished(QNetworkReply*, int, QStandardItem*)),
+             reply, task_id, item);
+}
+
+void RadioBrowserService::ShowContextMenu(const QPoint& global_pos) {
+  if (!model()->current_index().isValid()) return;
+  QStandardItem* item = model()->itemFromIndex(model()->current_index());
+  if (!item) return;
+
+  if (!context_menu_) {
+    context_menu_ = new QMenu;
+    context_menu_->addActions(GetPlaylistActions());
+    context_menu_->addAction(IconLoader::Load("download", IconLoader::Base),
+                             tr("Open %1 in browser").arg(homepage_url_.host()),
+                             this, SLOT(Homepage()));
+
+    context_menu_->addAction(IconLoader::Load("view-refresh", IconLoader::Base),
+                             tr("Refresh channels"), this,
+                             SLOT(ReloadSettings()));
+    context_menu_->addAction(IconLoader::Load("configure", IconLoader::Base),
+                             tr("Configure..."), this, SLOT(ShowConfig()));
+  }
+
+  context_menu_->popup(global_pos);
+}
```

Comment:
Note for future cleanup work: we could move the basic menu handling used in CloudFileService to the base class. I think each service implements this method, mostly with the same pattern.